jax/_src/callback.py

```diff
@@ -132,3 +133,103 @@ def _flat_callback(*flat_args):
+io_callback_p.def_impl(functools.partial(dispatch.apply_primitive,
+                                         io_callback_p))
+
+@io_callback_p.def_effectful_abstract_eval
+def io_callback_abstract_eval(*avals, callback: Callable[..., Any],
+                              result_avals, ordered: bool):
+  del avals, callback
+  effect = _OrderedIOEffect if ordered else _IOEffect
+  return result_avals, {effect}
+
+def io_callback_jvp_rule(*args, **kwargs):
+  del args, kwargs
+  raise ValueError("IO callbacks do not support JVP.")
+ad.primitive_jvps[io_callback_p] = io_callback_jvp_rule
+
+
+def io_callback_transpose_rule(*args, **kwargs):
+  del args, kwargs
+  raise ValueError("IO callbacks do not support transpose.")
+ad.primitive_transposes[io_callback_p] = io_callback_transpose_rule
+
+
+def io_callback_batching_rule(args, dims, callback, result_avals, ordered):
+  if ordered:
+    raise ValueError("Cannot `vmap` ordered IO callback.")
+  return pure_callback_batching_rule(args, dims, callback=callback,
+      vectorized=False, result_avals=result_avals)
+batching.primitive_batchers[io_callback_p] = io_callback_batching_rule
+
+def io_callback_lowering(ctx, *args, callback, ordered, **params):
+
+  def _callback(*flat_args):
+    return tuple(io_callback_impl(*flat_args, callback=callback,
+                                  ordered=ordered, **params))
+
+  if isinstance(ctx.module_context.axis_context,
+                (mlir.SPMDAxisContext, mlir.ShardingContext)):
+    # Apply maximal sharding so pjit only executes the callback on device 0.
+    sharding = xc.OpSharding()
+    sharding.type = xc.OpSharding.Type.MAXIMAL
+    sharding.tile_assignment_dimensions = [1]
+    sharding.tile_assignment_devices = [0]
+  else:
+    sharding = None
+
+  if ordered:
+    token = ctx.tokens_in.get(_OrderedIOEffect)[0]
+    result, token, keepalive = mlir.emit_python_callback(
+        ctx, _callback, token, list(args), ctx.avals_in, ctx.avals_out, True,
+        sharding=sharding)
+    ctx.set_tokens_out(mlir.TokenSet({_OrderedIOEffect: (token,)}))
+  else:
+    result, token, keepalive = mlir.emit_python_callback(
+        ctx, _callback, None, list(args), ctx.avals_in, ctx.avals_out, True,
+        sharding=sharding)
+  ctx.module_context.add_keepalive(keepalive)
+  return result
+mlir.register_lowering(io_callback_p, io_callback_lowering)
+
+def io_callback(callback: Callable[..., Any], result_shape_dtypes: Any,
+                *args: Any, ordered: bool = False, **kwargs: Any):
+  def _flat_callback(*flat_args):
+    args, kwargs = tree_util.tree_unflatten(in_tree, flat_args)
+    return tree_util.tree_leaves(callback(*args, **kwargs))
+
+  flat_args, in_tree = tree_util.tree_flatten((args, kwargs))
+  tree_util.tree_map(_check_shape_dtype, result_shape_dtypes)
+  result_avals = tree_util.tree_map(
+      lambda x: core.ShapedArray(x.shape, x.dtype), result_shape_dtypes)
+  flat_result_avals, out_tree = tree_util.tree_flatten(result_avals)
```

Comment:
Maybe flatten first here, so you can `map` rather than `tree_map` on the previous two checks